action/protocol/execution/evm/evmstatedbadapter.go

```diff
@@ -235,8 +235,26 @@ func (stateDB *StateDBAdapter) AddPreimage(common.Hash, []byte) {
 }
 
 // ForEachStorage loops each storage
-func (stateDB *StateDBAdapter) ForEachStorage(common.Address, func(common.Hash, common.Hash) bool) {
-	logger.Error().Msg("ForEachStorage is not implemented")
+func (stateDB *StateDBAdapter) ForEachStorage(addr common.Address, cb func(common.Hash, common.Hash) bool) {
+	ctt, err := stateDB.Contract(byteutil.BytesTo20B(addr[:]))
+	if err != nil {
+		// stateDB.err = err
+		return
+	}
+	iter, err := ctt.Iterator()
+	if err != nil {
+		// stateDB.err = err
+		return
+	}
+
+	for iter.Next() {
+		key, value := iter.Current()
```

Comment:
Is it the responsibility of iterator to guarantee  giving an immutable copy of k/v.